File: JAVA/GestionRelationEmployees/src/main/java/models/ProjectCollaborator.java
package models;

import java.sql.Date;
import java.sql.Timestamp;

/**
 * Modèle représentant un collaborateur assigné à un projet
 */
public class ProjectCollaborator {

    private int id;
    private int projectId;
    private int employeeId;
    private String role;  // Ex: "Chef de projet", "Développeur", "Designer"
    private Integer assignedHours;
    private int workedHours;
    private Date joinedDate;
    private Date leftDate;
    private boolean isActive;
    private Timestamp createdAt;

    // Champs transients pour l'affichage
    private String employeeName;
    private String projectName;

    // ─── Constructeurs ───────────────────────────────────────────────

    public ProjectCollaborator() {}

    public ProjectCollaborator(int projectId, int employeeId, String role, Integer assignedHours, Date joinedDate) {
        this.projectId = projectId;
        this.employeeId = employeeId;
        this.role = role;
        this.assignedHours = assignedHours;
        this.joinedDate = joinedDate;
        this.workedHours = 0;
        this.isActive = true;
    }

    // ─── Getters & Setters ───────────────────────────────────────────

    public int getId() { return id; }
    public void setId(int id) { this.id = id; }

    public int getProjectId() { return projectId; }
    public void setProjectId(int projectId) { this.projectId = projectId; }

    public int getEmployeeId() { return employeeId; }
    public void setEmployeeId(int employeeId) { this.employeeId = employeeId; }

    public String getRole() { return role; }
    public void setRole(String role) { this.role = role; }

    public Integer getAssignedHours() { return assignedHours; }
    public void setAssignedHours(Integer assignedHours) { this.assignedHours = assignedHours; }

    public int getWorkedHours() { return workedHours; }
    public void setWorkedHours(int workedHours) { this.workedHours = workedHours; }

    public Date getJoinedDate() { return joinedDate; }
    public void setJoinedDate(Date joinedDate) { this.joinedDate = joinedDate; }

    public Date getLeftDate() { return leftDate; }
    public void setLeftDate(Date leftDate) { this.leftDate = leftDate; }

    public boolean isActive() { return isActive; }
    public void setActive(boolean active) { isActive = active; }

    public Timestamp getCreatedAt() { return createdAt; }
    public void setCreatedAt(Timestamp createdAt) { this.createdAt = createdAt; }

    // Champs transients
    public String getEmployeeName() { return employeeName; }
    public void setEmployeeName(String employeeName) { this.employeeName = employeeName; }

    public String getProjectName() { return projectName; }
    public void setProjectName(String projectName) { this.projectName = projectName; }

    // ─── Méthodes utilitaires ────────────────────────────────────────

    /**
     * Retourne le taux d'avancement des heures travaillées
     */
    public String getHoursProgress() {
        if (assignedHours == null || assignedHours == 0) {
            return workedHours + "h";
        }
        return workedHours + "h / " + assignedHours + "h";
    }

    /**
     * Calcule le pourcentage d'heures utilisées
     */
    public double getHoursUsageRate() {
        if (assignedHours == null || assignedHours == 0) return 0;
        return (double) workedHours / assignedHours * 100;
    }

    /**
     * Vérifie si le collaborateur a dépassé ses heures allouées
     */
    public boolean isOverAllocated() {
        if (assignedHours == null) return false;
        return workedHours > assignedHours;
    }

    @Override
    public String toString() {
        return "ProjectCollaborator{id=" + id + ", employee=" + employeeName +
                ", role='" + role + "', hours=" + getHoursProgress() + "}";
    }
}

File: src/Entity/Projet/ProjectCollaborator.php
<?php

namespace App\Entity\Projet;

use App\Repository\Projet\ProjectCollaboratorRepository;
use Doctrine\DBAL\Types\Types;
use Doctrine\ORM\Mapping as ORM;

#[ORM\Entity(repositoryClass: ProjectCollaboratorRepository::class)]
#[ORM\Table(name: 'project_collaborators')]
#[ORM\Index(name: 'idx_project_collaborators_project_id', columns: ['project_id'])]
#[ORM\Index(name: 'idx_project_collaborators_employee_id', columns: ['employee_id'])]
#[ORM\Index(name: 'idx_project_collaborators_active', columns: ['is_active'])]
#[ORM\HasLifecycleCallbacks]
class ProjectCollaborator
{
    #[ORM\Id]
    #[ORM\GeneratedValue]
    #[ORM\Column]
    /** @phpstan-ignore-next-line */
    private ?int $id = null;

    #[ORM\Column(name: 'project_id', type: Types::INTEGER)]
    private ?int $projectId = null;

    #[ORM\Column(name: 'employee_id', type: Types::INTEGER)]
    private ?int $employeeId = null;

    #[ORM\Column(length: 120)]
    private ?string $role = null;

    #[ORM\Column(name: 'assigned_hours', type: Types::INTEGER, nullable: true)]
    private ?int $assignedHours = null;

    #[ORM\Column(name: 'worked_hours', type: Types::INTEGER)]
    private int $workedHours = 0;

    #[ORM\Column(name: 'joined_date', type: Types::DATE_MUTABLE)]
    private ?\DateTimeInterface $joinedDate = null;

    #[ORM\Column(name: 'left_date', type: Types::DATE_MUTABLE, nullable: true)]
    private ?\DateTimeInterface $leftDate = null;

    #[ORM\Column(name: 'is_active', type: Types::BOOLEAN)]
    private bool $isActive = true;

    #[ORM\Column(name: 'created_at', type: Types::DATETIME_MUTABLE, nullable: true)]
    private ?\DateTimeInterface $createdAt = null;

    public function getId(): ?int
    {
        return $this->id;
    }

    public function getProjectId(): ?int
    {
        return $this->projectId;
    }

    public function setProjectId(int $projectId): static
    {
        $this->projectId = $projectId;
        return $this;
    }

    public function getEmployeeId(): ?int
    {
        return $this->employeeId;
    }

    public function setEmployeeId(int $employeeId): static
    {
        $this->employeeId = $employeeId;
        return $this;
    }

    public function getRole(): ?string
    {
        return $this->role;
    }

    public function setRole(string $role): static
    {
        $this->role = $role;
        return $this;
    }

    public function getAssignedHours(): ?int
    {
        return $this->assignedHours;
    }

    public function setAssignedHours(?int $assignedHours): static
    {
        $this->assignedHours = $assignedHours;
        return $this;
    }

    public function getWorkedHours(): int
    {
        return $this->workedHours;
    }

    public function setWorkedHours(int $workedHours): static
    {
        $this->workedHours = $workedHours;
        return $this;
    }

    public function getJoinedDate(): ?\DateTimeInterface
    {
        return $this->joinedDate;
    }

    public function setJoinedDate(\DateTimeInterface $joinedDate): static
    {
        $this->joinedDate = $joinedDate;
        return $this;
    }

    public function getLeftDate(): ?\DateTimeInterface
    {
        return $this->leftDate;
    }

    public function setLeftDate(?\DateTimeInterface $leftDate): static
    {
        $this->leftDate = $leftDate;
        return $this;
    }

    public function isActive(): bool
    {
        return $this->isActive;
    }

    public function setIsActive(bool $isActive): static
    {
        $this->isActive = $isActive;
        return $this;
    }

    public function getCreatedAt(): ?\DateTimeInterface
    {
        return $this->createdAt;
    }

    #[ORM\PrePersist]
    public function onPrePersist(): void
    {
        $this->createdAt = new \DateTime();
    }
}

